Revisit how the Android port works.

paulrouget opened the pull request:
Fix #20855 
Fix #18625 
Fix #21147 

Before polishing and making sure everything works fine (like the VR code, the android-x86 version, the non-android version of the lib, …), I'd like to get some early feedback on the approach.

I recommend reviewing commit by commit.

To test, just follow the regular steps:

```
    ./mach build -d --android
    ./mach package --dev --android builds servo.apk
    ./mach install --dev --android && ./mach run --android
```

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20912)
<!-- Reviewable:end -->

paulrouget committed:
new android port: Remove Android code from /ports/servo/

cbrewster commented on ports/libsimpleservo/README.md:
```diff
@@ -0,0 +1,3 @@
+This is a basic wrapper around Servo. While libservo itself (/components/servo/) offers a lot of flexibility,
+libsimpleservo (/ports/libsimpleservo/) tries to make it easier to embed Servo, without much configuration needed.
+It is limited to only one view (no tabs, no multiple renering area).
```
s/renering area/rendering areas/

jdm commented on ports/libsimpleservo/build.rs:
```diff
@@ -0,0 +1,21 @@
+use std::path::Path;
+
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
```
```
Note that if the build script itself (or one of its dependencies) changes, then it's rebuilt and rerun unconditionally, so cargo:rerun-if-changed=build.rs is almost always redundant (unless you want to ignore changes in all other files except for build.rs).
```
From https://doc.rust-lang.org/cargo/reference/build-scripts.html

jdm commented on ports/libsimpleservo/src/api.rs:
```diff
@@ -0,0 +1,338 @@
+
+struct ResourceReader(Box<ReadFileTrait>);
+unsafe impl Send for ResourceReader {}
+unsafe impl Sync for ResourceReader {}
```
This is scary. What if we store `Box<ReadFileTrait + Send + Sync>` instead?

jdm reviewed (commented):
This looks like a really useful improvement :D

paulrouget commented:
WebVR build segfaults:

```
06-19 03:17:48.509 17357-17599/com.mozilla.servo A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0xc in tid 17599 (WebVRThread), pid 17357 (m.mozilla.servo)
```

paulrouget commented:
@MortimerGoro I might need your help here.

I get a crash in `self.service.get_displays()` (components/webvr/webvr_thread.rs) which is a rust-webvr method. I'm building with `--features googlevr`. It works well on master, but with this current PR it segfaults.

This PR changes how Servo runs on Android. Instead of using an empty Android app and load Servo via Glutin, we have a proper Android app that use a GLSurfaceView and load servo as a library and make some calls via JNI. I don't know what `get_displays` does, do you have any hint of what could go wrong? When can get_displays crash?

In the meantime, I will compile with my own version of rust-webvr see if I can find precisely where it crashes.

@jdm / @MortimerGoro also, in general, do you know how to get a rust stack trace on Android when it crashes?

MortimerGoro commented:
@paulrouget I suspect that the crash is related to JNI calls which get the activity this way: android::get_app().activity

Can check check that the activity has a correct value. If not we'll need to implement a different way to get it or expose a method in rust-webvr to set it from servo.

MortimerGoro commented:
@paulrouget have you check if WebGL works ok? IIRC GLSurfaceView was not a very friendly API to create shared contexts

paulrouget commented:
@MortimerGoro WebGL works.

paulrouget commented:
@MortimerGoro where is `android::get_app().activity` called?

paulrouget commented:
Ok, found it.

paulrouget commented:
@MortimerGoro I don't use android rs glue. Should I? Does it even make sense for a library?

If I don't, what's the alternative? We could add some embedder methods to access the activity. Would that be enough?

MortimerGoro commented:
I think that the best way would be to get rid of android-rs-glue, and pass the JNI JavaVM and Activity pointers to rust-webvr (e.g. in the constructor or initialization methods).

jdm commented:
To answer the question about stack traces, I do the following:
* copy the logcat output that includes the crash information and unresolved symbols to a temporary file `log`
* run `ndk-stack -sym target/armv7-linux-androideabi/debug/apk/jniLibs/armeabi-v7a/ -dump log`
(it might be necessary to pass target/armv7-linux-androideabi/debug/apk/obj/local/armeabi-v7a/lib/ for the sym directory instead; I forget which is correct at the moment)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20421) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@MortimerGoro what if I initialize the android glue manually (calling `main2`)?

paulrouget commented:
> what if I initialize the android glue manually (calling main2) by setting ANDROID_APP manually?

I tried, but struggling. My approach is, in Java, get the application via:

```java
mApp = ((Activity) context).getApplication();
// mApp is not null
```

passed to Rust:

```rust
init(.., app: JObject, ..) {
    unsafe {
        android_injected_glue::ANDROID_APP = app.into_inner() as *mut android_injected_glue::ffi::android_app;
    }
    // get_app().activity crashes
}
```

MortimerGoro commented:
@paulrouget You can't convert the java Application instance into the native android_app struct. They are different things. android_app struct is an ad-hoc struct created specially for NativeActivity (which glutin uses).

If you want to initialize the android glue manually I'd follow this steps:
- Allocate the android_app struct
- Reset all the struct to 0
- Manually assign the following values: 
  - ANativeWindow* window; (you may need to use this helper function: ANativeWindow_fromSurface(..)
  - ANativeActivity struct (allocate & reset to 0 too)
    - JavaVM* vm
    - JNIEnv* env
    - jobject clazz (this is the java activity instance, use jni->NewGlobalRef(activity) instead of direct assign to allow it to be used from any thread

Check this code for the original android_app struct definition used in the NDK

[native_app_glue.zip](https://github.com/servo/servo/files/2118474/native_app_glue.zip)

paulrouget commented:
@MortimerGoro Understood. This helps.

paulrouget commented:
Note: I need to migrate from the JNI create to android_injected_glue. It's quite annoying because the JNI crate comes with few helpers I rely on. But the JNI crate doesn't expose enough to build the objects needed by the android_injected_glue, and it's becoming messy to have to JNI crates.

I also tried to remove the android_injected_glue dependency from rust-webvr, but it's quite some work.

paulrouget commented:
This is what my code looks like. All I did so far is to bind vm and the jobject of the activity.

The problem I have is that I use the `jni-rs` and `rust-jni-sys`, while `webvr` uses `android_injected_glue`.

I'd rather keep `jni-rs` if possible, so I tried to create the `ANDROID_APP` object from the raw JNI pointers held by the jnr-rs wrappers. But I hit a `null pointer dereference` in `JNIScope::attach` in webvr.

Any idea if what I'm doing could actually work?

Maybe I'm trying to hard to keep jnirs :)

```rust

pub fn Java_com_mozilla_servoview_NativeServo_init(env: JNIEnv, _: JClass, activity: JObject) {

    // Thread is GLThread

    // From jni-rs to android_injected_glue
    let vm /* *mut *const JNIInvokeInterface from jni crate */ = env.get_java_vm().unwrap().get_native_interface();
    let vm = unsafe {
        (*vm) as *const android_injected_glue::ffi::JNIInvokeInterface
    };

    let clazz:/* *mut enum {} */ jobject = env.new_global_ref(activity).unwrap().as_obj().into_inner();
    let clazz = clazz as *mut std::os::raw::c_void;

    let mut native_activity: android_injected_glue::ffi::ANativeActivity = unsafe { std::mem::zeroed() };
    native_activity.vm = &mut android_injected_glue::ffi::JavaVM { functions: vm };
    native_activity.clazz = clazz;

    let mut app: android_injected_glue::ffi::android_app = unsafe { std::mem::zeroed() };
    app.activity = &native_activity;

    unsafe { // use of mutable static
        android_injected_glue::ANDROID_APP = &mut app;
    }

```

MortimerGoro commented:
@paulrouget IIRC I discarted JNI crate because it didn't find Java classes when used in threads created from Rust (which don't have a classloader attached and you need to get it from the activity as a I did).

paulrouget commented:
The latest commit fixes webvr.

Depends on https://github.com/prevoty/jni-rs/pull/98

All blockers are fixed now. There's still some code missing for full non-android support, and exporting the Surface and the View in a AAR, but I will do that in a followup PR.

I will polish this PR, file the follow-ups, and then ask for a review.

paulrouget commented:
To file once landed:

- support starting Android app with custom URL (support command line arguments)
- Re-introduce pinch to zoom
- make /ports/servo/ use libsimpleservo
- support prefs: `shell.searchpage / shell.homepage / shell.keep_screen_on.enabled / shell.native-orientation`
- support OpenGL
- properly set `enable_subpixel_text_antialiasing` for VR
- CI builds of libsimpleservo on non android platforms
- build an example out of the c-api

paulrouget commented:
@MortimerGoro Can I ask you to look at the code change in /supports/ and at ports/libsimpleservo/src/jniapi.rs ?

@SimonSapin Do you mind looking at the /python/ changes? I had to tweak the build scripts a bit to get this new port to compile properly

SimonSapin commented on python/servo/build_commands.py:
```diff
@@ -231,6 +235,7 @@ def build(self, target=None, release=False, dev=False, jobs=None,
 
         if android:
             target = self.config["android"]["target"]
+            opts += ["-p", "libsimpleservo"]
```
Rather than this change, I’d prefer to have libsimpleservo in `default-members` in the workspace’s `Cargo.toml`.

(If that crate doesn’t compile on some platforms use `#[cfg]` so that it does, possibly as `#![cfg]` to disable the whole crate.)

paulrouget replied:
Done.

SimonSapin commented on python/servo/build_commands.py:
```diff
@@ -412,4 +419,4 @@ def clean(self, manifest_path=None, params=[], verbose=False):
         opts += params
         return check_call(["cargo", "clean"] + opts,
-                          env=self.build_env(), cwd=self.servo_crate(), verbose=verbose)
+                          env=self.build_env(), cwd=self.servo_crate(android), verbose=verbose)
```
I don’t know if this is needed for `clean`, since `cwd` mostly affects optional feature selection.

SimonSapin commented on python/servo/post_build_commands.py:
```diff
@@ -82,15 +82,14 @@ def run(self, params, release=False, dev=False, android=None, debug=False, debug
-                    % param.replace("'", "\\'")
-                ]
+            extra = ""
+            # FIXME: see params comment above. This is supposed to be parsed by servo's `from_cmdline_args`.
```
Not supporting arguments other than the URL is a significant regression. I’m considering relying on `--webdriver` to run WPT.

paulrouget replied:
Done.

SimonSapin commented on ports/libsimpleservo/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+
+[features]
+googlevr = ["libservo/googlevr"]
+oculusvr = ["libservo/oculusvr"]
```
You probably also want:

```toml
default = ["unstable"]
unstable = ["libservo/unstable"]
```

… and maybe others from `ports/servo/Cargo.toml`.

paulrouget replied:
Done.

paulrouget committed:
new android port: minimal Android app that relies on libsimpleservo

paulrouget committed:
new android port: update fakeld

paulrouget commented:
@SimonSapin The changes in /python/servo/ are more sparse now. And all the parameters are now supported.

SimonSapin commented:
Thanks. r+ on `/python/servo` changes.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21094) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21087) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget committed:
new android port: introduce a simple servo library

MortimerGoro commented:
r+ on /supports/ and at ports/libsimpleservo/src/jniapi.rs

Great work!

paulrouget commented:
I would use some help:

Appveyor: Build execution time has reached the maximum allowed time for your plan (90 minutes).

Travis: LLVM ERROR: IO failure on output stream: No space left on device

paulrouget commented:
@bors-servo retry

paulrouget commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit c0a51523f56da9d48c464a056f6e878ae59a32a1 with merge 5823b41ecadd9b84d2c077c1903a969cb06cc086...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/11838)

paulrouget commented:
@SimonSapin 

What should be done about this:

http://build.servo.org/builders/android/builds/11838/steps/compile__1/logs/stdio

```
You have not accepted the license agreements of the following SDK components:
  [Solver for ConstraintLayout 1.0.0, ConstraintLayout for Android 1.0.0].
  Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
  Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
```

SimonSapin commented:
This is not using the new `./mach bootstrap-package` command, so this SDK’s license should have been accepted here: https://github.com/servo/saltfs/blob/90036502c8fac2291fba372b3309bafef605f5b2/servo-build-dependencies/android.sls#L92

paulrouget commented:
@SimonSapin I'm not sure to understand. What is `./mach bootstrap-package`? You mean `bootstrap-android`?

It's for a specific component: ConstraintLayout 1.0.0.

SimonSapin commented:
Ugh yes I did. Was that component not used before?

paulrouget commented:
> Was that component not used before?

Yes. Introduced in this PR.

paulrouget commented:
Current state of things:

- Appveyor: Build execution time has reached the maximum allowed time for your plan (90 minutes).
- Travis: LLVM ERROR: IO failure on output stream: No space left on device
- homu: You have not accepted the license agreements of the following SDK components:  [Solver for ConstraintLayout 1.0.0, ConstraintLayout for Android 1.0.0].

I could use some help.

/cc @SimonSapin for the homu error, @jdm for the travis and appveyor errors.

For the 90 minutes error, I'm wondering if it actually takes more time to compile now. Maybe because of the feature disparities we end up compiling twice.

jdm commented:
Yes, test-unit is definitely recompiling a large number of crates all over again, which would explain the problems with timing out (and likely why we're running out of disk space on travis).

paulrouget commented:
appveyor is passing, but it took 88 minutes. Still too long. Looking at the logs, it's still compiling more than usual.

paulrouget commented:
Test-unit takes 35 minutes to compile on Windows, after PR, instead of 9 minutes. Investigating the extra compilations.

paulrouget commented:
Before PR:

```
mach test-unit
   Compiling servo_remutex v0.0.1 (file:///C:/projects/servo/components/remutex)
   Compiling selectors v0.19.0 (file:///C:/projects/servo/components/selectors)
   Compiling msg v0.0.1 (file:///C:/projects/servo/components/msg)
   Compiling embedder_traits v0.0.1 (file:///C:/projects/servo/components/embedder_traits)
   Compiling servo_config v0.0.1 (file:///C:/projects/servo/components/config)
   Compiling bluetooth_traits v0.0.1 (file:///C:/projects/servo/components/bluetooth_traits)
   Compiling canvas_traits v0.0.1 (file:///C:/projects/servo/components/canvas_traits)
   Compiling profile_traits v0.0.1 (file:///C:/projects/servo/components/profile_traits)
   Compiling net_traits v0.0.1 (file:///C:/projects/servo/components/net_traits)
   Compiling style v0.0.1 (file:///C:/projects/servo/components/style)
   Compiling profile v0.0.1 (file:///C:/projects/servo/components/profile)
   Compiling net v0.0.1 (file:///C:/projects/servo/components/net)
   Compiling script_traits v0.0.1 (file:///C:/projects/servo/components/script_traits)
   Compiling profile_tests v0.0.1 (file:///C:/projects/servo/tests/unit/profile)
   Compiling gfx v0.0.1 (file:///C:/projects/servo/components/gfx)
   Compiling style_tests v0.0.1 (file:///C:/projects/servo/tests/unit/style)
   Compiling metrics v0.0.1 (file:///C:/projects/servo/components/metrics)
   Compiling script_layout_interface v0.0.1 (file:///C:/projects/servo/components/script_layout_interface)
   Compiling script v0.0.1 (file:///C:/projects/servo/components/script)
   Compiling layout v0.0.1 (file:///C:/projects/servo/components/layout)
   Compiling metrics_tests v0.0.1 (file:///C:/projects/servo/tests/unit/metrics)
   Compiling script_tests v0.0.1 (file:///C:/projects/servo/tests/unit/script)
    Finished dev [unoptimized + debuginfo] target(s) in 9m 12s
```

After PR:

```
mach test-unit
   Compiling winapi v0.3.5
   Compiling log v0.4.1
   Compiling log v0.3.9
   Compiling servo_remutex v0.0.1 (file:///C:/projects/servo/components/remutex)
   Compiling gl_generator v0.9.0
   Compiling mime v0.2.4
   Compiling gleam v0.5.1
   Compiling offscreen_gl_context v0.17.0
   Compiling euclid v0.17.3
   Compiling rand v0.4.2
   Compiling wincolor v0.1.6
   Compiling libloading v0.5.0
   Compiling same-file v1.0.2
   Compiling rand v0.3.22
   Compiling uuid v0.6.2
   Compiling rayon-core v1.4.0
   Compiling parking_lot_core v0.2.13
   Compiling termcolor v0.3.6
   Compiling clang-sys v0.23.0
   Compiling walkdir v2.1.4
   Compiling hyper v0.10.13
   Compiling phf_generator v0.7.21
   Compiling ipc-channel v0.10.0
   Compiling servo_rand v0.0.1 (file:///C:/projects/servo/components/rand)
   Compiling rayon v1.0.0
   Compiling parking_lot v0.5.4
   Compiling env_logger v0.5.10
   Compiling style v0.0.1 (file:///C:/projects/servo/components/style)
   Compiling hyper_serde v0.8.0
   Compiling hyper-openssl v0.2.6
   Compiling influent v0.4.0
   Compiling servo-websocket v0.21.0
   Compiling string_cache_codegen v0.4.1
   Compiling phf_codegen v0.7.21
   Compiling webrender_api v0.57.2 (https://github.com/servo/webrender#db26bc0e)
   Compiling jpeg-decoder v0.1.14
   Compiling bindgen v0.37.4
   Compiling selectors v0.19.0 (file:///C:/projects/servo/components/selectors)
   Compiling cssparser-macros v0.3.3
   Compiling mime_guess v1.8.1
   Compiling script v0.0.1 (file:///C:/projects/servo/components/script)
   Compiling string_cache v0.7.3
   Compiling markup5ever v0.7.1
   Compiling servo_atoms v0.0.1 (file:///C:/projects/servo/components/atoms)
   Compiling image v0.18.0
   Compiling mozjs_sys v0.51.4
   Compiling cssparser v0.24.0
   Compiling xml5ever v0.12.0
   Compiling html5ever v0.22.3
   Compiling mozjs v0.7.1
   Compiling malloc_size_of v0.0.1 (file:///C:/projects/servo/components/malloc_size_of)
   Compiling servo_url v0.0.1 (file:///C:/projects/servo/components/url)
   Compiling msg v0.0.1 (file:///C:/projects/servo/components/msg)
   Compiling range v0.0.1 (file:///C:/projects/servo/components/range)
   Compiling malloc_size_of_tests v0.0.1 (file:///C:/projects/servo/tests/unit/malloc_size_of)
   Compiling style_traits v0.0.1 (file:///C:/projects/servo/components/style_traits)
   Compiling gfx_traits v0.0.1 (file:///C:/projects/servo/components/gfx_traits)
   Compiling webvr_traits v0.0.1 (file:///C:/projects/servo/components/webvr_traits)
   Compiling devtools_traits v0.0.1 (file:///C:/projects/servo/components/devtools_traits)
   Compiling embedder_traits v0.0.1 (file:///C:/projects/servo/components/embedder_traits)
   Compiling servo_geometry v0.0.1 (file:///C:/projects/servo/components/geometry)
   Compiling servo_config v0.0.1 (file:///C:/projects/servo/components/config)
   Compiling bluetooth_traits v0.0.1 (file:///C:/projects/servo/components/bluetooth_traits)
   Compiling net_traits v0.0.1 (file:///C:/projects/servo/components/net_traits)
   Compiling profile_traits v0.0.1 (file:///C:/projects/servo/components/profile_traits)
   Compiling canvas_traits v0.0.1 (file:///C:/projects/servo/components/canvas_traits)
   Compiling profile v0.0.1 (file:///C:/projects/servo/components/profile)
   Compiling net v0.0.1 (file:///C:/projects/servo/components/net)
   Compiling script_traits v0.0.1 (file:///C:/projects/servo/components/script_traits)
   Compiling profile_tests v0.0.1 (file:///C:/projects/servo/tests/unit/profile)
   Compiling gfx v0.0.1 (file:///C:/projects/servo/components/gfx)
   Compiling style_tests v0.0.1 (file:///C:/projects/servo/tests/unit/style)
   Compiling metrics v0.0.1 (file:///C:/projects/servo/components/metrics)
   Compiling script_layout_interface v0.0.1 (file:///C:/projects/servo/components/script_layout_interface)
   Compiling layout v0.0.1 (file:///C:/projects/servo/components/layout)
   Compiling metrics_tests v0.0.1 (file:///C:/projects/servo/tests/unit/metrics)
   Compiling script_tests v0.0.1 (file:///C:/projects/servo/tests/unit/script)
    Finished dev [unoptimized + debuginfo] target(s) in 34m 40s
     Running target\debug\deps\font_context-52b2747725e6ef20.exe
```

paulrouget commented:
It was because we were not using ports/servo/Cargo.toml for the tests.

paulrouget commented:
Now, the proper amount of crates are compiled. But still taking too long (probably would have taken 50% more time, it got canceled almost at the end).

As for travis, we moved from 28.6G to 30.6G (tested on my linux machine).

paulrouget commented:
Looking at the appveyor logs, the only extra crates compiled are `libsimpleservo`, `enum-iterator`, and ` enum-iterator-derive`. But still, both `build` and `test-unit` took 50% more time to compile.

Looking at the rustc calls, they are exactly the same.

paulrouget commented:
Travis issue: libsimpleservo, before being stripped, is 1.6Gb and located in 2 places: target/debug/deps/libsimpleservo.so and target/debug/libsimpleservo.so. That explains the extra space.

paulrouget commented:
Travis has 11.9 Gb available (before rustup and downloading crates). On my computer, /target/ hits precisely 11.9 Gb after `mach unit-test`. 11Gb on master. We reached the travis limit (target + rustup + crates).

See also: https://docs.travis-ci.com/user/reference/overview/#Virtualisation-Environment-vs-Operating-System

I suggest that we don't compile libsimpleservo on Linux. Only Windows and Mac.

@jdm: what do you think?

paulrouget commented:
I'm not 100% sure about the last commit. I'm trying to build libsimpleservo only on Mac and Windows. @jdm does it look right to you?

SimonSapin commented:
Notes for future Cargo debugging:

`RUST_LOG=cargo ./mach build 2> /tmp/log` gets the full log, which is very verbose. A lot happens before compilation starts, so the command can be killed after a few seconds.

`RUST_LOG=cargo::core::resolver` filters specifically for dependency graph resolution, such as why a given version was selected over another. It’s still rather verbose.

`RUST_LOG=cargo::core::resolver ./mach test-unit 2>&1 | sed -n '/features:/,$p' | sed '/DEBUG/,$d'>/tmp/log` cuts further for only keeping the list of feature flags enabled for each crate.

`RUST_LOG=cargo::core::compiler::fingerprint` can help figure out why something is being recompiled. When the output looks like `fingerprint error for ipc-channel v0.10.0: failed to read …/target/debug/.fingerprint/ipc-channel-c8c57f5a7b6dfa50/… cause: No such file or directory`, it’s usually because the hash computed by Cargo (`c8c57f5a7b6dfa50` in this case) is different from the previous build. However this doesn’t say why: the hash includes a number of things like the rustc version, the compilation profile, the extra compiler flags, …

paulrouget commented:
Ok. It's now green.

Got a review for the build system changes, for the java code.

@jdm r? for commits:

- Remove Android code from /ports/servo/
- introduce a simple servo library
- only build libsimpleservo on Mac and Windows

paulrouget committed:
Update Gradle and SDK

jdm commented on ports/libsimpleservo/build.rs:
```diff
@@ -0,0 +1,125 @@
+        (target.clone(), target)
+    };
+
+    let toolchain_path = ndk_path.join("toolchains").join(format!("{}-4.9", toolchain)).join("prebuilt").
```
Rather than building these up manually, it would be

jdm commented on python/servo/command_base.py:
```diff
@@ -330,15 +330,26 @@ def get_target_dir(self):
+        if release:
+            return path.join(base_path, "release", apk_name)
+        else:
+            return path.join(base_path, "debug", apk_name)
```
build_type = "release" if release else "debug"
return path.join(base_path, build_type, apk_name)

paulrouget replied:
Done.

jdm commented on python/servo/package_commands.py:
```diff
@@ -421,7 +421,7 @@ def install(self, release=False, dev=False, android=False, emulator=False, usb=F
 
         if android:
-            pkg_path = binary_path + ".apk"
+            pkg_path = self.get_apk_path(release)
```
Hmm, does this fix #20148?

paulrouget replied:
No.

paulrouget commented on ports/libsimpleservo/build.rs:
```diff
@@ -0,0 +1,125 @@
+    let directory = Path::new(&out_dir);
+
+    // compiling android_native_app_glue.c
+    if Command::new(toolchain_path.join("bin").join(format!("{}-gcc", prefix)))
```
Done.

paulrouget commented on ports/libsimpleservo/build.rs:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    // compiling libandroid_native_app_glue.a
+    if Command::new(toolchain_path.join("bin").join(format!("{}-ar", prefix)))
```
Done.

paulrouget commented on ports/libsimpleservo/src/api.rs:
```diff
@@ -0,0 +1,455 @@
+        info!("go_back");
+        let event = WindowEvent::Navigation(self.browser_id, TraversalDirection::Back(1));
+        self.events.push(event);
+        if !self.batch_mode {
```
Done.

paulrouget commented on ports/libsimpleservo/src/capi.rs:
```diff
@@ -0,0 +1,260 @@
+
+#[no_mangle]
+pub extern "C" fn servo_version() -> *const c_char {
+    let v = api::servo_version();
```
Done.

paulrouget commented on ports/libsimpleservo/src/jniapi.rs:
```diff
@@ -0,0 +1,385 @@
+}
+
+#[no_mangle]
+#[allow(non_snake_case)]
```
Done.

jdm commented:
Looks good! Squash in whatever way makes sense!

paulrouget commented:
@jdm done.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 1a33357 has been approved by `jdm`

<!-- @bors-servo r=jdm 1a33357b64fef576779a7d80a58c6d5bedf69532 -->

bors-servo commented:
:hourglass: Testing commit 1a33357b64fef576779a7d80a58c6d5bedf69532 with merge b0c5a45e6a264d62e81dff0511c0bd420d84c798...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/11892)

jdm commented:
```
File /home/servo/.android/repositories.cfg could not be loaded.
Checking the license for package Solver for ConstraintLayout 1.0.0 in /home/servo/android/sdk/r25.2.3/licenses
Warning: License for package Solver for ConstraintLayout 1.0.0 not accepted.
Checking the license for package ConstraintLayout for Android 1.0.0 in /home/servo/android/sdk/r25.2.3/licenses
Warning: License for package ConstraintLayout for Android 1.0.0 not accepted.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> You have not accepted the license agreements of the following SDK components:
  [Solver for ConstraintLayout 1.0.0, ConstraintLayout for Android 1.0.0].
  Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
  Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

paulrouget commented:
Apparently, there is an issue with the early versions of constraintlayout and license/auto-download. I think we are basically hitting this issue: https://stackoverflow.com/questions/42731625/travis-ci-failed-because-cannot-accept-license-constrain-layout . People suggest to either copy the license files or to add specific instructions to accept the constraintlayout license.

But my hope is that the SDK update (#21160) will fix this issue. I already got a r+ from @MortimerGoro , so if the SDK update fixes this issue, I will just include the SDK update to this PR.

paulrouget commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 52c91d5 has been approved by `jdm`

<!-- @bors-servo r=jdm 52c91d545b83a2ce9082aef7c1eed51326551445 -->

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/11895)

CYBAI commented:
```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':servoapp'.
> Failed to install the following Android SDK packages as some licences have not been accepted.
     platforms;android-27 Android SDK Platform 27
     build-tools;27.0.3 Android SDK Build-Tools 27.0.3
  To build this project, accept the SDK license agreements and install the missing components using the Android Studio SDK Manager.
  Alternatively, to transfer the license agreements from one workstation to another, see http://d.android.com/r/studio-ui/export-licenses.html
  
  Using Android SDK: /home/servo/android/sdk/r25.2.3

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 26s
```

jdm commented:
As mentioned in the other PR, I believe we need some salt changes to update the SDK.

paulrouget commented:
/cc @SimonSapin can you help me with this? I'm not sure what needs to be done here. See the "update SDK" commit.

paulrouget commented:
@jdm ok - I'll look.

paulrouget commented:
Saltfs PR: https://github.com/servo/saltfs/pull/856

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21195) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21200) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21213) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo r=jdm,mortimergoro,simonsapin

bors-servo commented:
:pushpin: Commit 5b926be has been approved by `jdm,mortimergoro,simonsapin`

<!-- @bors-servo r=jdm,mortimergoro,simonsapin 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 -->

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/12011)

jdm commented:
From the script that checks for certain symbols in the resulting binary:
```
Traceback (most recent call last):
  File "./etc/ci/check_dynamic_symbols.py", line 23, in <module>
    os.environ['ANDROID_NDK'], 'toolchains', 'arm-linux-androideabi-4.9',
  File "/usr/lib/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'ANDROID_NDK'
program finished with exit code 1
```

bors-servo commented:
:hourglass: Testing commit 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 with merge a3284daff3e258e3e518373437026d30a199b736...

bors-servo commented:
:hourglass: Testing commit 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 with merge 4ccae586f2b4b1853a7f9e757ecaf68258ab9391...

bors-servo commented:
:hourglass: Testing commit 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 with merge 98b3cac32d2c5c53f5b65a905e5652e58a5aaaad...

bors-servo commented:
:hourglass: Testing commit 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 with merge d404d698f9c03deea103dd5af0b154065d76ef92...

jdm commented:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PerformanceResourceTiming to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1

  ▶ Unexpected subtest result in /_mozilla/mozilla/window_performance.html:
  │ FAIL [expected PASS] window_performance
  │   → PerformanceTiming is not defined
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/window_performance.html:14:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/window_performance.html:9:1

 ▶ Unexpected subtest result in /hr-time/timeOrigin.html:
  └ PASS [expected FAIL] Window and worker timeOrigins differ when worker is created after a delay.

  ▶ Unexpected subtest result in /hr-time/window-worker-timeOrigin.window.html:
  └ PASS [expected FAIL] timeOrigin and now() should be correctly ordered between window and worker

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker.html:
  │ FAIL [expected PASS] interfaces
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PerformanceResourceTiming to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.worker.js:9:1

  ▶ OK [expected CRASH] /workers/worker-performance.worker.html

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] Resource timing seems to work in workers
  │   → assert_equals: getEntriesByType("resource") returns one entry with initiatorType of xmlhttprequest expected 1 but got 0
  │ 
  │ testPerformanceResourceTiming@http://web-platform.test:8000/workers/worker-performance.worker.js:70:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:62:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] Can use performance.getEntriesByType in workers
  │   → assert_equals: Measure entryhas the right duration expected 0.06765100000000146 but got 0
  │ 
  │ verifyEntry@http://web-platform.test:8000/workers/worker-performance.worker.js:10:5
  │ testPerformanceGetEntriesByType@http://web-platform.test:8000/workers/worker-performance.worker.js:46:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:39:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] performance.timing is not available in workers
  │   → assert_equals: performance.timing is undefined expected "undefined" but got "function"
  │ 
  │ testPerformanceHasNoTiming@http://web-platform.test:8000/workers/worker-performance.worker.js:107:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:106:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] Can use clearMarks and clearMeasures in workers
  │   → assert_equals: clearMarks cleared the marks expected 0 but got 2
  │ 
  │ testPerformanceClearing@http://web-platform.test:8000/workers/worker-performance.worker.js:57:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:55:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] performance.setResourceTimingBufferSize in workers
  │   → performance.setResourceTimingBufferSize is not a function
  │ 
  │ testPerformanceSetResourceTimingBufferSize@http://web-platform.test:8000/workers/worker-performance.worker.js:99:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:98:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] performance.clearResourceTimings in workers
  │   → performance.clearResourceTimings is not a function
  │ 
  │ testPerformanceClearResourceTimings@http://web-platform.test:8000/workers/worker-performance.worker.js:94:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:93:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] performance.toJSON is available in workers
  │   → assert_equals: performance.toJSON is a function expected "function" but got "undefined"
  │ 
  │ testPerformanceHasToJSON@http://web-platform.test:8000/workers/worker-performance.worker.js:115:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:114:1

  ▶ Unexpected subtest result in /workers/worker-performance.worker.html:
  │ FAIL [expected PASS] Performance marks and measures seem to be working correctly in workers
  │   → assert_equals: measure's startTime is the first mark's startTime expected 11.776454 but got 0
  │ 
  │ testPerformanceEntryOrder@http://web-platform.test:8000/workers/worker-performance.worker.js:52:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ @http://web-platform.test:8000/workers/worker-performance.worker.js:49:1

  ▶ Unexpected subtest result in /_mozilla/mozilla/window_performance_topLevelDomComplete.html:
  │ FAIL [expected PASS] performance.topLevelDomComplete
  │   → assert_true: expected true got false
  │ 
  │ window.onload<@http://web-platform.test:8000/_mozilla/mozilla/window_performance_topLevelDomComplete.html:11:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1562:20

  ▶ Unexpected subtest result in /hr-time/test_cross_frame_start.html:
  │ FAIL [expected PASS] Child and parent time bases are correct
  │   → assert_true: Child's now() is based on its document's navigationStart expected true got false
  │ 
  │ finish_test/<@http://web-platform.test:8000/hr-time/test_cross_frame_start.html:46:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ finish_test@http://web-platform.test:8000/hr-time/test_cross_frame_start.html:39:15

  ▶ Unexpected subtest result in /hr-time/test_cross_frame_start.html:
  │ FAIL [expected PASS] Child created at least 1 second after parent
  │   → assert_true: expected true got false
  │ 
  │ finish_test/<@http://web-platform.test:8000/hr-time/test_cross_frame_start.html:36:34
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1538:20
  │ test@http://web-platform.test:8000/resources/testharness.js:545:9
  └ finish_test@http://web-platform.test:8000/hr-time/test_cross_frame_start.html:36:15
```

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4796)

jdm commented:
```
  = note: Undefined symbols for architecture x86_64:
            "std::__1::__tree<sh::ImageFunctionHLSL::ImageFunction, std::__1::less<sh::ImageFunctionHLSL::ImageFunction>, std::__1::allocator<sh::ImageFunctionHLSL::ImageFunction> >::value_comp()", referenced from:
                std::__1::__tree_node_base<void*>*& std::__1::__tree<sh::ImageFunctionHLSL::ImageFunction, std::__1::less<sh::ImageFunctionHLSL::ImageFunction>, std::__1::allocator<sh::ImageFunctionHLSL::ImageFunction> >::__find_equal<sh::ImageFunctionHLSL::ImageFunction>(std::__1::__tree_node_base<void*>*&, sh::ImageFunctionHLSL::ImageFunction const&) in libmozangle-7a2c668cb5c47e04.rlib(ImageFunctionHLSL.o)
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
That's... new?

jdm commented:
The earlier android failure is actually https://github.com/servo/saltfs/issues/869, not the fault of these changes.

jdm commented:
@bors-servo retry
* checking if it's just servo-mac3 being weird again

bors-servo commented:
:hourglass: Testing commit 5b926be2f914aa4bc4a4fc9fe4ff22319d2cb6b3 with merge 25caf1e8a5640b67884c1b63ff5442f741156b02...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/12026)

CYBAI commented:
```
android-toolchains/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-objdump: 'target/armv7-linux-androideabi/debug/libservo.so': No such file
Traceback (most recent call last):
  File "./etc/ci/check_dynamic_symbols.py", line 26, in <module>
    'target/armv7-linux-androideabi/debug/libservo.so']
  File "/usr/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['android-toolchains/ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-objdump', '-T', 'target/armv7-linux-androideabi/debug/libservo.so']' returned non-zero exit status 1
```

meecoffee commented:
why so many failed！

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21242) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
Just missing a rename in python script. Will rebase and make the change.

paulrouget commented:
@bors-servo r=jdm,mortimergoro,simonsapin

bors-servo commented:
:pushpin: Commit 0e1cabf has been approved by `jdm,mortimergoro,simonsapin`

<!-- @bors-servo r=jdm,mortimergoro,simonsapin 0e1cabf05a2c38f9d2c15650784a82c28ff22caa -->

bors-servo commented:
:hourglass: Testing commit 0e1cabf05a2c38f9d2c15650784a82c28ff22caa with merge 15aab833d99f8031fe8b18efd0d311f845a70445...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9083)

jdm commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/12040), [arm32](http://build.servo.org/builders/arm32/builds/12247), [arm64](http://build.servo.org/builders/arm64/builds/12292), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12338), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9254), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9258), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12384), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4808), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4817), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9122), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4892), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4833), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7740) are reusable. Rebuilding only [android-x86](http://build.servo.org/builders/android-x86/builds/158), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9083)...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/159)

paulrouget commented:
I can reproduce:

```
Assuming --target i686-linux-android
emulator: Requested console port 5580: Inferring adb port 5581.
emulator: Cold boot: requested by the user
emulator: INFO: boot completed
Success
Traceback (most recent call last):
  File "/home/paul/git/servo/etc/run_in_headless_android_emulator.py", line 212, in <module>
    sys.exit(main(*sys.argv[1:]))
  File "/home/paul/git/servo/etc/run_in_headless_android_emulator.py", line 61, in main
    check_call(adb + ["shell", cmd], stdout=sys.stderr)
  File "/home/paul/git/servo/etc/run_in_headless_android_emulator.py", line 123, in check_call
    exit_code = call(*args, **kwargs)
  File "/home/paul/git/servo/etc/run_in_headless_android_emulator.py", line 118, in call
    with terminate_on_exit(*args, **kwargs) as process:
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
emulator: Discarding the changed state: command-line flag
emulator: WARNING: Discarding the changed state (command-line flag).
    return self.gen.next()
  File "/home/paul/git/servo/etc/run_in_headless_android_emulator.py", line 95, in terminate_on_exit
    process = subprocess.Popen(*args, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 390, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1025, in _execute_child
    raise child_exception
TypeError: execv() arg 2 must contain only strings
EOF without finding the expected line
```

SimonSapin commented:
Guess based on only looking at the trackback: `cmd` in `check_call(adb + ["shell", cmd], stdout=sys.stderr)` might be a Unicode string, but a byte string is needed.

SimonSapin commented on etc/run_in_headless_android_emulator.py:
```diff
@@ -55,10 +55,10 @@ def main(avd_name, apk_path, *args):
 
-        check_call(adb + ["shell", "am start com.mozilla.servo/com.mozilla.servo.MainActivity"],
-                   stdout=sys.stderr)
+        cmd_args = "-e servoargs \'" + " ".join(args) + "\'" if args else ""
```
Joining with spaces like this breaks the case where one of the argument contains a space. Could we use one of these instead? (From `adb shell am -h`)

```
    [--esa <EXTRA_KEY> <EXTRA_STRING_VALUE>[,<EXTRA_STRING_VALUE...]]
        (mutiple extras passed as String[]; to embed a comma into a string,
         escape it using "\,")
    [--esal <EXTRA_KEY> <EXTRA_STRING_VALUE>[,<EXTRA_STRING_VALUE...]]
        (mutiple extras passed as List<String>; to embed a comma into a string,
         escape it using "\,")
```

And `shell_quote` should still be used.

SimonSapin commented on ports/libsimpleservo/src/api.rs:
```diff
@@ -0,0 +1,418 @@
+
+    // opts::from_cmdline_args expects the first argument to be the binary name.
+    let argsline = format!("servo {}", argsline);
+    let args: Vec<String> = argsline.split_whitespace().map(|s| s.to_owned()).collect();
```
`split_whitespace` accepts a number of characters, including some non-ASCII, as separators. This should at least be `.split(' ')`, but see other comment.

SimonSapin commented:
@bors-servo r-

I’d prefer the argument escaping to be fixed before landing, but if you feel strongly this is better left to a follow-up feel free to r+ again.

paulrouget commented:
We are going to run into a timeout as we are waiting on a logcat message that will never arrive because the new port doesn't redirect stdout to logcat.

There are few ways to do that:

- make `setprop log.redirect-stdio true` work
- do like injected glue and copy stdout via `__android_log_write`
- use log macros that are already redirected to logcat
- as mentioned by Simon, use the undocumented std::io::set_print function that redirects `println!()` https://github.com/rust-lang/rust/blob/master/src/libstd/io/stdio.rs#L638-L659
- delegate console.log printing to embedder

Relevant: #19999

For now, I think the best thing to do is to use `alert` instead of `log` for the test. This is already delegated to the embedder so it's a very easy fix. And file a follow-up to redirect stdout to logcat.

I'll look at this tomorrow.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21250) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
I managed to solve the logcat issue, but now, apparently, the servo event loop is not working in headless mode. I'm trying to figure out what's going on.

paulrouget commented:
@SimonSapin addressed your comments. Using JSON now.

Also, I made it so that `window.alert` redirector the logcat, fixing the initial problem. It's not ideal, I'd like console.log to do the same thing, but we do not delegate console.log to the embedder. Not a problem for now, but we will need to fix this: https://github.com/servo/servo/issues/19999#issuecomment-409228907

paulrouget committed:
new android port: fix android build and tests

paulrouget commented:
@bors-servo r=jdm,mortimergoro,simonsapin

jdm commented:
@bors-servo r=jdm,mortimergoro,simonsapin

bors-servo commented:
:pushpin: Commit a923e8b has been approved by `jdm,mortimergoro,simonsapin`

<!-- @bors-servo r=jdm,mortimergoro,simonsapin a923e8b5911144b6e7ab6fba4c3d8efb09f99cb8 -->

jdm commented:
@bors-servo r=jdm,mortimergoro,simonsapin

bors-servo commented:
:pushpin: Commit 6a31864 has been approved by `jdm,mortimergoro,simonsapin`

<!-- @bors-servo r=jdm,mortimergoro,simonsapin 6a31864126679486935bd80c5f7f1c8ef7b027ed -->

bors-servo commented:
:hourglass: Testing commit 6a31864126679486935bd80c5f7f1c8ef7b027ed with merge 95b54ca8f3c931e1902bcafd517e9466a9a0f8af...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12069), [android-x86](http://build.servo.org/builders/android-x86/builds/184), [arm32](http://build.servo.org/builders/arm32/builds/12272), [arm64](http://build.servo.org/builders/arm64/builds/12316), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12364), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9281), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9284), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12408), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4837), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4843), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9111), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9149), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4917), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4859), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7765)
Approved by: jdm,mortimergoro,simonsapin
Pushing 95b54ca8f3c931e1902bcafd517e9466a9a0f8af to master...

Darkspirit commented:
This PR deleted occurrences of `android_params`.
Should apparently unused file `/resources/android_params` now be deleted, too?

paulrouget commented:
> This PR deleted occurrences of `android_params`.
> Should apparently unused file `/resources/android_params` now be deleted, too?

Likely. Can you file a dedicated issue for that?
